Content/Items/Moonstone/Weapons.CrescentQuarterstaff.cs

```diff
@@ -49,507 +58,661 @@ public override void AddRecipes()
+				for (int i = 0; i < 16; i++)
 				{
-					moveDirection.X = -moveDirection.X;
-					Projectile.ai[1] = 0f;
+					cache.Add(StaffEnd);
 				}
 			}
 
-			Projectile.velocity = speed * moveDirection;
-			Projectile.velocity = Collide();
+			for (int i = 0; i < 16; i++)
+			{
+				cache[i] = cache[i] - Vector2.UnitY * 8;
+			}
+
+			cache.Add(StaffEnd);
+
+			while (cache.Count > 16)
+			{
+				cache.RemoveAt(0);
+			}
+		}
+
+		private void ManageTrail()
+		{
+			trail ??= new Trail(Main.instance.GraphicsDevice, 16, new TriangularTip(8), factor => 50f * Charge, factor => new Color(78, 87, 191) * Charge * 0.7f * factor.X);
+
+			trail.Positions = cache.ToArray();
+
+			trail.NextPosition = (StaffEnd - Player.MountedCenter).RotatedBy(Player.direction - Math.PI / 6) + Player.MountedCenter;
 		}
 
-		protected virtual Vector2 Collide()
+		public void DrawPrimitives()
 		{
-			return Collision.noSlopeCollision(Projectile.position, Projectile.velocity, Projectile.width, Projectile.height, true, true);
+			if (CurrentAttack == AttackType.Slam && slamCharged)
+			{
+				Main.spriteBatch.End();
+
+				Effect effect = Filters.Scene["CeirosRing"].GetShader().Shader;
+
+				var world = Matrix.CreateTranslation(-Main.screenPosition.Vec3());
+				Matrix view = Main.GameViewMatrix.ZoomMatrix;
+				var projection = Matrix.CreateOrthographicOffCenter(0, Main.screenWidth, Main.screenHeight, 0, -1, 1);
+
+				effect.Parameters["time"].SetValue(Main.GameUpdateCount * 0.02f);
+				effect.Parameters["repeats"].SetValue(1f);
+				effect.Parameters["transformMatrix"].SetValue(world * view * projection);
+				effect.Parameters["sampleTexture"].SetValue(Request<Texture2D>("StarlightRiver/Assets/EnergyTrail").Value);
+
+				trail?.Render(effect);
+
+				Main.spriteBatch.Begin(default, default, default, default, default, default, Main.GameViewMatrix.TransformationMatrix);
+			}
 		}
 
 		public void DrawAdditive(SpriteBatch spriteBatch)
 		{
+			if (CurrentAttack == AttackType.Slam && slamCharged)
+			{
+				Texture2D texFlare = ModContent.Request<Texture2D>(AssetDirectory.Assets + "StarTexture").Value;
+				Texture2D texBloom = ModContent.Request<Texture2D>(AssetDirectory.Keys + "GlowAlpha").Value;
+
+				float flareRotation = MathHelper.SmoothStep(0, MathHelper.TwoPi, timer / 40f);
+				float flareScale = timer < 20 ? MathHelper.SmoothStep(0, 1, timer / 20f) : MathHelper.SmoothStep(1, 0, (timer - 20) / 20f);
+
+				var color = new Color(78, 87, 191);
+
+				Vector2 pos = StaffEnd + Vector2.UnitX.RotatedBy(Projectile.rotation) * 10 * Projectile.scale;
+
+				spriteBatch.Draw(texBloom, pos - Main.screenPosition, null, color, 0, texBloom.Size() / 2, Projectile.scale * 3 * flareScale, default, default);
+				spriteBatch.Draw(texFlare, pos - Main.screenPosition, null, color * 2, flareRotation, texFlare.Size() / 2, Projectile.scale * 0.75f * flareScale, default, default);
+			}
+		}
+	}
+
+	public class CrescentOrb : ModProjectile, IDrawAdditive
```

Comment:
While typically we try to keep all projectiles for one item together, the large size of this file may mean it is a good decision to pull this class out to a seperate file